lib/widgets/message_list.dart

```diff
@@ -901,23 +902,29 @@ class _MarkAsReadAnimationState extends State<MarkAsReadAnimation> {
 class RecipientHeader extends StatelessWidget {
   const RecipientHeader({super.key, required this.message, required this.narrow});
 
-  final Message message;
+  final MessageBase message;
   final Narrow narrow;
 
   @override
   Widget build(BuildContext context) {
     final message = this.message;
-    return switch (message) {
-      StreamMessage() => StreamMessageRecipientHeader(message: message, narrow: narrow),
-      DmMessage() => DmRecipientHeader(message: message, narrow: narrow),
-    };
+    switch (message) {
+      case MessageBase<StreamConversation>():
+        return StreamMessageRecipientHeader(message: message, narrow: narrow);
+      case MessageBase<DmConversation>():
+        return DmRecipientHeader(message: message, narrow: narrow);
+      case MessageBase<Conversation>():
+        assert(false, 'Unexpected concrete subclass of MessageBase<Conversation>:'
+                      ' ${objectRuntimeType(message, 'MessageBase<Conversation>')}');
+        return SizedBox.shrink();
```

Comment:
I think we can handle this with more confidence:
```suggestion
      case MessageBase<Conversation>():
        throw StateError('Bad concrete subclass of MessageBase');
```

We're not going to be writing lots more of these subclasses, so there won't be many opportunities to get this wrong.

That also lets this go back to a switch-expression.